gloo-timers: don't consume callback in Interval closure

samcday opened the pull request:
Otherwise Interval only works on first callback and fails after that

OddCoincidence commented on crates/timers/src/callback.rs:
```diff
@@ -154,7 +154,7 @@ impl Interval {
         let mut callback = Some(callback);
         let closure = Closure::wrap(Box::new(move || {
-            let mut callback = callback.take().unwrap_throw();
+            let callback = callback.as_mut().take().unwrap_throw();
```
I think this could be simplified to:
```
let closure = Closure::wrap(callback);`
```

Pauan replied:
Because of `take()`, I'm pretty sure this will still break.

So I agree with @OddCoincidence, that this should just completely remove the `Some` wrapper and use `Closure::wrap(Box::new(callback) as Box<FnMut()>)` instead

samcday replied:
Yeah I already updated my local copy to `Closure::wrap(Box::new(callback) as Box<FnMut()>)`, I'm just grappling with the integration tests at the moment and should have the changes pushed up soon.

samcday replied:
Also the current version with `.take()` does work and is re-entrant. It works because `as_mut` is creating a new Option that is a mutable reference. So `take()` on a `.as_mut()` is not consuming the original Option.

OddCoincidence commented:
Could we also add a test that triggers this bug?

samcday commented:
@OddCoincidence yep working on tests now. As it happens integration tests are not running in CI, which is why it's reporting that these changes are okay here, when in fact fixing the bug actually causes the `interval_cancel` test to fail.

samcday commented:
@OddCoincidence alright I wrangled the tests. Hope you don't mind but I did a little bit of re-jiggery once I understood how they worked to simplify them a little.

The `interval` test now ensures that 5 ticks of the interval pass. `interval_stream` was already doing that so now they're consistent. I removed the `Cell` stuff in favour of just `collect()`ing the Stream and counting how many results we got.

The `interval_cancel` test started failing once I fixed the original bug. I reworked it a little to avoid the dance it was doing in favour of just combining a Timeout stream with the `Interval` stream we're examining.

fitzgen commented on crates/timers/tests/web.rs:
```diff
@@ -91,53 +91,57 @@ fn timeout_future_cancel() -> impl Future<Item = (), Error = wasm_bindgen::JsVal
+            sender.try_send(()).unwrap()
+        }
+    })
+    .forget();
```
We shouldn't forget this -- that will cause the `setInterval` call to keep running even after the test is done.

Instead, we should do `let interval = Interval::new(...)` and then in the `and_then` closure, we should `drop(interval)` so that it gets cleared once the test is finished.

samcday replied:
Ah yes, good call. Done!

samcday commented:
Sweet, feedback addressed. Looks like CI is running headless tests now too, nice!
